src/buffer.rs

```diff
@@ -0,0 +1,113 @@
+use std::borrow::{Borrow, Cow};
+use std::io::Read;
+use std::ops::Deref;
+use std::hash::{Hash, Hasher};
+
+use bytes::{Buf, Bytes};
+use string;
+
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub(crate) struct StrBuf(string::String<Bytes>);
+
+impl StrBuf {
+    pub(crate) fn from_str(s: &str) -> Self {
+        StrBuf(string::String::from_str(s))
+    }
+
+    pub(crate) fn from_utf8(bytes: ByteBuf) -> Result<Self, ::std::str::Utf8Error> {
+        Ok(StrBuf(string::TryFrom::try_from(bytes.0)?))
+    }
+
+    pub(crate) fn from_utf8_lossy(bytes: ByteBuf) -> Self {
+        match String::from_utf8_lossy(bytes.as_ref()) {
+            Cow::Owned(s) => s.into(),
+            Cow::Borrowed(s) => {
+                // SAFETY: We know that `bytes` only contains valid utf-8,
+                //         since the `from_utf8_lossy` operation returned the
+                //         input verbatim.
+                debug_assert_eq!(s.len(), bytes.len());
+                StrBuf(unsafe { string::String::from_utf8_unchecked(bytes.clone().0) })
+            }
+        }
+    }
+}
+
+impl From<String> for StrBuf {
+    fn from(s: String) -> Self {
+        let ByteBuf(bytes) = ByteBuf::from(s.into_bytes());
```

Comment:
Why not create the `bytes::Bytes` directly from `s.into_bytes()` with `.into()`, instead of round-tripping it through `ByteBuf::from` ? (Though it would get optimized the same anyway.)

Reply:
Good catch, this must be a leftover of a refactor I did. I'll get rid of it!